File: chap7/enc.c
#include <stdio.h>
#include <string.h>
#include <err.h>

#define PRIME (17)
#define SHIFT (11)

/**
 * @plain: input text
 * @shift: an integer
 * @prime: a prime number
 */
char* ceaser(const char *plain, int shift, int prime) {
  const int len = strlen(plain);
  char *cipher = (char *)malloc(len);
  
  for (int i = 0 ; i < len ; i ++) {
    *(cipher + i) = (*(plain + i) - 'a' + shift) % prime + 'a';
  }
  *(cipher + len) = '\0';

  return cipher;
}

int main(int argc, char *argv[]) {
  if (argc != 2) {
    err(1, "usage: %s [text]\n", argv[0]);
  }

  const char *encrypted = ceaser(argv[1], SHIFT, PRIME);
  const char *decrypted = ceaser(encrypted, -SHIFT, PRIME);

  printf("'%s' =enc->> '%s' =dec->> '%s'\n",
	 argv[1], encrypted, decrypted);
  
  return 0;
}

File: chap9/enc.py
#!/usr/bin/python

import os
import sys

def ceaser(plain, shift, prime):
    def _group(c):
        return chr((ord(c) - ord("a") + shift) % prime + ord("a"))
    return "".join(map(_group, plain))

if __init__ == "__main__":
    PRIME = 17
    SHIFT = 11

    if len(sys.argv) != 2:
        print("usage: %s [text]" % sys.argv[0])
        exit(1)

    txt = sys.argv[1]
    enc = ceaser(txt,  SHIFT, PRIME)
    dec = ceaser(enc, -SHIFT, PRIME)

    print("'%s' =enc->> '%s' =dec->> '%s'\n" % (txt, enc, dec))
